frb_codegen/src/parser/mod.rs

```diff
@@ -213,6 +215,137 @@ fn extract_comments(attrs: &[Attribute]) -> Vec<IrComment> {
         .collect()
 }
 
+pub mod frb_keyword {
+    syn::custom_keyword!(mirror);
+    syn::custom_keyword!(non_final);
+    syn::custom_keyword!(metadata);
+    syn::custom_keyword!(import);
+}
+
+#[derive(Clone, Debug)]
+pub struct NamedOption<K, V> {
+    pub name: K,
+    pub value: V,
+}
+
+impl<K: Parse + std::fmt::Debug, V: Parse> Parse for NamedOption<K, V> {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let name: K = input.parse()?;
+        let _: Token![=] = input.parse()?;
+        let value = input.parse()?;
+        Ok(Self { name, value })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct MirrorOption(Path);
+
+impl Parse for MirrorOption {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let content;
+        parenthesized!(content in input);
+        let path: Path = content.parse()?;
+        Ok(Self(path))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct MetadataAnnotations(Vec<IrAnnotation>);
+
+impl Parse for IrAnnotation {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let annotation: LitStr = input.parse()?;
+        let library = if input.peek(frb_keyword::import) {
+            let _ = input.parse::<frb_keyword::import>()?;
+            let library: IrImport = input.parse()?;
+            Some(library)
+        } else {
+            None
+        };
+        Ok(Self {
+            content: annotation.value(),
+            library: library,
+        })
+    }
+}
+impl Parse for MetadataAnnotations {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let content;
+        parenthesized!(content in input);
+        let annotations = Punctuated::<IrAnnotation, syn::Token![,]>::parse_terminated(&content)?
+            .into_iter()
+            .collect();
+        Ok(Self(annotations))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DartImports(Vec<IrImport>);
+
+impl Parse for IrImport {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let uri: LitStr = input.parse()?;
+        let alias: Option<String> = if input.peek(token::As) {
+            let _ = input.parse::<token::As>()?;
+            let alias: Ident = input.parse()?;
+            Some(alias.to_string())
+        } else {
+            None
+        };
+        Ok(Self {
+            uri: uri.value(),
+            alias: alias,
+        })
+    }
+}
+impl Parse for DartImports {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let content;
+        parenthesized!(content in input);
+        let imports = Punctuated::<IrImport, syn::Token![,]>::parse_terminated(&content)?
+            .into_iter()
+            .collect();
+        Ok(Self(imports))
+    }
+}
+
+enum FrbOption {
+    Mirror(MirrorOption),
+    NonFinal,
+    Metadata(NamedOption<frb_keyword::metadata, MetadataAnnotations>),
+}
```

Comment:
If I understand correctly, you successfully parsed *all* rust macros? That sounds great!

Seems that, the mirror/nonfinal/... already have their own parsers to parse them. Thus are we duplicating works here? Maybe we can refactor those?

Reply:
Yes, I try to make an uniform way to parse parameters in "[frb]" macros. I think we can refactor code to use it. But I am new here, so I think it's better to move things slowly and do it later. ^_^

Reply:
Sure. Look forward to your future PRs!